Command that gives the player resistance to specific types of damage.

EaguaireSama opened the pull request:
This possibly solves the issue #2837

- **Commands**

**Command Name-damageResist**
Parameters - (damagetype) (percentage)
         Description - **damageResist** gives the player resistance to any type of damage 
**Bonus Command-CheckResistance**
Parameters - null
         Description - **CheckResistance** returns the percentage values of each damage type so the player can keep track of the values.

 - **Note**
1. **ALL** can be used (instead of the damagetype) to change the percentage values of all the types in one line.
2. the player can use **damageResist ALL 100** to gain Global Damage Immunity.

EaguaireSama committed:
First commit

e-aakash commented on src/main/java/org/terasology/logic/health/DamageResistComponent.java:
```diff
@@ -0,0 +1,52 @@
+ * limitations under the License.
+ */
+public class DamageResistComponent implements Component {
+   private int value;
```
What is the use of this variable?

EaguaireSama replied:
Oh no,that variable is redundant..I must have missed it while I changed the component class..
I can remove it and request another pr if you want

e-aakash replied:
No need to send new pr, just make changes and push to the same branch. This PR will get updated.

EaguaireSama replied:
wow I did not know that.will do.

e-aakash commented on src/main/java/org/terasology/logic/health/DamageResistComponent.java:
```diff
@@ -0,0 +1,52 @@
+   Map<String,Integer> damageTypes;
+
+    public DamageResistComponent() {
+    damageTypes=new HashMap<>(100);
```
Is an initialization of 100 required?

EaguaireSama replied:
I just had to be sure there's no collision.

e-aakash reviewed (commented):
Haven't tested yet. 
Also, improved commit messages to reflect the changes made would be better.

EaguaireSama committed:
removed redundant attributes and also did some minor improvements

EaguaireSama committed:
changed checkResistance to work on multiplayer

syntaxi commented on src/main/java/org/terasology/logic/health/DamageResistComponent.java:
```diff
@@ -0,0 +1,39 @@
+    damageTypes=new HashMap<>(6);
+    this.setAll(0);
+    }
+    public void setAll(int d){
```
Part of the Entity-Component-System structure is that there should be no logic in the components. Instead the recommended method is to have the logic in systems instead.

So for instance, here, I would replace the whole component with simply:
```java
public Map<String, Integer> damageTypes;
```
and then move this `setAll` method into the command system.

EaguaireSama replied:
Sure,I will take care of that.

syntaxi commented on src/main/java/org/terasology/logic/health/DamageResistEventHandler.java:
```diff
@@ -0,0 +1,48 @@
+public class DamageResistEventHandler extends BaseComponentSystem {
+
+    @ReceiveEvent(components = {DamageResistComponent.class},priority = EventPriority.PRIORITY_HIGH)
+    public void suitingOn(BeforeDamagedEvent event, EntityRef entity) {
```
All event handlers should have javadoc :)

Also, this is just incase you were not aware, you can move the `components` bit of
```java
@ReceiveEvent(components = {DamageResistComponent.class},priority = EventPriority.PRIORITY_HIGH)
```
into the method signature itself, like so:
```java
public void suitingOn(BeforeDamagedEvent event, EntityRef entity, DamageResistComponent resistanceComponent) {
  // Method body
}
```

This means you can then remove the line:
```java
DamageResistComponent resistance=entity.getComponent(DamageResistComponent.class);
```
as well as making the inclusion of the component clearer, as well as what you are doing with it.

EaguaireSama replied:
Oh I guess I didn't read the Wiki well enough 😃 .I will change that part and write the Javadoc as well.

syntaxi commented on src/main/java/org/terasology/logic/health/HealthCommands.java:
```diff
@@ -72,6 +72,51 @@ public String damageWithType(@Sender EntityRef client, @CommandParam("damageType
+                }
+            }
+
+            if(damageResist.damageTypes.containsKey(dType)){
```
This is fine for this PR, but once this PR is merged I'd like to see an issue opened about making this more dynamic as to allow for damage types beyond the few listed in the `setAll` method. That way, if a module decided to say add "psychic" as a damage type then it would be automatically handled by the system.

Like I said, not something that needs to be changed now, but I just wanted to make a note about it :)

EaguaireSama replied:
Yeah sure.And maybe If I have time I'll look into it myself. 👍

syntaxi commented on src/main/java/org/terasology/logic/health/DamageResistEventHandler.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.ArrayList;
+import java.util.List;
+
+import java.sql.SQLOutput;
```
Why is this an import? Probably a left over from IDE weirdness I assume?

EaguaireSama replied:
yeah probably :D

syntaxi reviewed (changes requested):
Overall, very nice.

A couple of nitpicks but they're not very big and should be easily resolved. Once fixed I'm happy with the changes.

As a aside note, if you're feeling up to it, I'd love to see an addition whereby if the player doesn't specify a percentage t will default to 100%.
Ie, `damageResist ALL` would give 100% resistance to all damages.

EaguaireSama committed:
changes have been made as per the requests

EaguaireSama commented:
I'm sorry for my late response..I couldn't find a fix for one of the changes...
But, I did create another command called **damageImmune** to solve it.
**damageImmune (damageType)** takes 100% as the default percentage.
I know this is not the best way to do it so I'd love some suggestions.

e-aakash reviewed (approved):
I am happy with the current state of this PR. 👍 

I would like some documentation (at least what are the commands and what they do) somewhere in the repo- either in README or as an article in wiki. This can be done in separate PR.

EaguaireSama commented:
Sure,will do.

Cervator commented on src/main/java/org/terasology/logic/health/DamageAuthoritySystem.java:
```diff
@@ -236,4 +236,18 @@ public void onCrash(HorizontalCollisionEvent event, EntityRef entity, CharacterS
+     */
+    @ReceiveEvent
+    public void onResistMode(BeforeDamagedEvent event, EntityRef entity, DamageResistComponent resistanceComponent) {
+        String damageType=event.getDamageType().getName();
```
You've got some minor code style issues in various files here, mostly missing whitespace, like wrapping spaces around `=` :-)

Helps to grab a Checkstyle plugin for your IDE and the warnings will highlight automatically.

Cervator commented on src/main/java/org/terasology/logic/health/DamageAuthoritySystem.java:
```diff
@@ -236,4 +236,18 @@ public void onCrash(HorizontalCollisionEvent event, EntityRef entity, CharacterS
+    @ReceiveEvent
+    public void onResistMode(BeforeDamagedEvent event, EntityRef entity, DamageResistComponent resistanceComponent) {
+        String damageType=event.getDamageType().getName();
+        String subString=damageType.substring(7);
```
Be wary of using substring with a hard coded number - the value may change or leave a viewer wondering what it means ("magic numbers") - could lead to bugs later :-)

You may want to:

* Find a way to dynamically find the right spot (like using RegEx)
* Alternatively extract the number into a namked constant to reduce confusion
* Harden the code in case if a failure to find expected values (look for an index of a desired char, if not found then log an error / throw an exception, etc)

Cervator commented on src/main/java/org/terasology/logic/health/DamageAuthoritySystem.java:
```diff
@@ -236,4 +236,18 @@ public void onCrash(HorizontalCollisionEvent event, EntityRef entity, CharacterS
+    public void onResistMode(BeforeDamagedEvent event, EntityRef entity, DamageResistComponent resistanceComponent) {
+        String damageType=event.getDamageType().getName();
+        String subString=damageType.substring(7);
+        float data=100-resistanceComponent.damageTypes.get(subString);
```
`100` here again may be a magic number, I figure you're calculating a percentage - you may also want to consider the impact of using integers and floats together in math (possible loss of precision), Finally you risk errors by assuming several levels of data are all going to be valid and found :-)

Cervator commented on src/main/java/org/terasology/logic/health/HealthCommands.java:
```diff
@@ -72,6 +74,73 @@ public String damageWithType(@Sender EntityRef client, @CommandParam("damageType
     }
+    public void setAll(DamageResistComponent damageResist,int v){
+        damageResist.damageTypes=new HashMap<>(6);
+        damageResist.damageTypes.put("directDamage",v);
```
Somewhat akin to what Jelly already said, but a slightly different angle and still maybe not for this PR: rather than in-lining the values here it may be wise to introduce a set of common damage types constants somewhere (if we don't already have those?) and use those instead here. Still as plain Strings, not like references or enums or anything, but just to avoid defining the values in multiple places.

Cervator commented on src/main/java/org/terasology/logic/health/HealthCommands.java:
```diff
@@ -72,6 +74,73 @@ public String damageWithType(@Sender EntityRef client, @CommandParam("damageType
+        if(value>=0&&value<=100) {
+            ClientComponent player = clientEntity.getComponent(ClientComponent.class);
+            DamageResistComponent damageResist = player.character.getComponent(DamageResistComponent.class);
+            String s="ALL";
```
Another example of missing spaces here and elsewhere, also try to avoid single character variable names - make them express more meaning :-)

Cervator commented on src/main/java/org/terasology/logic/health/HealthCommands.java:
```diff
@@ -72,6 +74,73 @@ public String damageWithType(@Sender EntityRef client, @CommandParam("damageType
+                damageResist = player.character.getComponent(DamageResistComponent.class);
+                setAll(damageResist,0);
+            }
+            if(dType.equals(s)){
```
Odd abbreviations are also discouraged - IDEs make longer names easy anyway 👍

Cervator commented on src/main/java/org/terasology/logic/health/HealthCommands.java:
```diff
@@ -72,6 +74,73 @@ public String damageWithType(@Sender EntityRef client, @CommandParam("damageType
+        ClientComponent player= clientEntity.getComponent(ClientComponent.class);
+        DamageResistComponent damageResist= player.character.getComponent(DamageResistComponent.class);
+
+        if(damageResist==null){
```
Rather than do a null check you can use a `hasComponent` check and then either indicate the target has no resistances if not or then retrieve the component knowing it exists

Cervator reviewed (changes requested):
Various minor style and code quality suggestions. Nothing too critical but also lots of easy tweaks for better code 👍

EaguaireSama committed:
removed the setall method and made the mapping dynamic,improved the code quality

EaguaireSama commented:
@Cervator I removed the ***setAll*** method and changed the code a little bit. Now the damage types are not put into the map at the beginning but instead are put according to the player's choosing. So, even if someone decides to add another damage type to the engine there won't be a problem.
And furthermore ***checkResistance*** now only shows the ***damageTypes*** that are currently active.

DarkWeird commented on src/main/java/org/terasology/logic/health/DamageAuthoritySystem.java:
```diff
@@ -236,4 +236,29 @@ public void onCrash(HorizontalCollisionEvent event, EntityRef entity, CharacterS
+        String subString = damageType.substring(damageType.indexOf(':') + 1);
+        float amount;
+        String key;
+        if(resistanceComponent.damageTypes.containsKey("all")||resistanceComponent.damageTypes.containsKey(subString)){
```
minor style issue: add spaces around ```||``` - more readable

DarkWeird commented on src/main/java/org/terasology/logic/health/DamageAuthoritySystem.java:
```diff
@@ -236,4 +236,29 @@ public void onCrash(HorizontalCollisionEvent event, EntityRef entity, CharacterS
+        String damageType = event.getDamageType().getName();
+        //takes the damage type name from the prefab name
+        String subString = damageType.substring(damageType.indexOf(':') + 1);
+        float amount;
```
minor style issue: Not necessary to declare variable at this place - looks , you can write data in variable at more then one place

DarkWeird commented on src/main/java/org/terasology/logic/health/HealthCommands.java:
```diff
@@ -15,6 +15,8 @@
  */
 package org.terasology.logic.health;
 
+import com.sun.xml.internal.ws.api.pipe.Engine;
```
minor style issue: not used import, using jvm internal api("com.sun.*")

DarkWeird commented on src/main/java/org/terasology/logic/health/HealthCommands.java:
```diff
@@ -15,6 +15,8 @@
 package org.terasology.logic.health;
 
+import com.sun.xml.internal.ws.api.pipe.Engine;
+import org.mockito.internal.matchers.Null;
```
minor style issue: not used import, using test dependencies in "production" code

DarkWeird commented on src/main/java/org/terasology/logic/health/HealthCommands.java:
```diff
@@ -72,6 +77,63 @@ public String damageWithType(@Sender EntityRef client, @CommandParam("damageType
+                return "Resistance:" + value + "% to " + damageType;
+            }
+        else{ return "accepted values:[0 to 100]";
+        }
```
many small style issue:
Normalize bracket wraps plz.
after ```if (...)```  - ```<space>{```
"else" wraps as ```} else {```
code on new line after "{"

I don't sure about tera's rules about one-line if body... possibly it must warped with brackets.

(You can use IDEA formatter for this all (Windows: Ctrl+Alt+L))

EaguaireSama committed:
resolved style issues
